File: tests/test_expirations.py
"""
test_expirations.py - Test item TTL expiration behavior.
"""

import os
import sys
import time
import unittest

sys.path.insert(0, os.path.dirname(__file__))
from memcached_harness import MemcachedServer


class TestExpirations(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.server = MemcachedServer().start()

    @classmethod
    def tearDownClass(cls):
        cls.server.stop()

    def setUp(self):
        self.mc = self.server.client()

    def tearDown(self):
        self.mc.close()

    def test_ttl_expiration(self):
        # Set with 2-second TTL
        self.assertEqual(self.mc.set("expire_key", "expire_val", exptime=2), "STORED")
        res = self.mc.get("expire_key")
        self.assertIn("expire_key", res)

        # Wait 2.5s for expiration
        time.sleep(2.5)

        # Should now be gone
        res = self.mc.get("expire_key")
        self.assertNotIn("expire_key", res)

    def test_permanent_item(self):
        # Exptime 0 means never expire
        self.assertEqual(self.mc.set("perm_key", "perm_val", exptime=0), "STORED")
        time.sleep(1)
        res = self.mc.get("perm_key")
        self.assertIn("perm_key", res)


if __name__ == "__main__":
    unittest.main()

File: tests/test_expirations.c
/*
 * test_expirations.c - Test item TTL expiration behavior.
 */

#include "c_harness.h"

static int test_ttl_expiration(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  char resp[64];
  char val[64];

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(client_set(&c, "expire_key", "expire_val", 10, 0, 2, resp,
                         sizeof(resp)) == 0);
  TEST_ASSERT_STR_EQ(resp, "STORED");

  TEST_ASSERT(client_get(&c, "expire_key", val, sizeof(val), NULL) == 0);
  TEST_ASSERT_STR_EQ(val, "expire_val");

  test_sleep_ms(2500);

  TEST_ASSERT(client_get(&c, "expire_key", val, sizeof(val), NULL) != 0);

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

static int test_permanent_item(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  char resp[64];
  char val[64];

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(
      client_set(&c, "perm_key", "perm_val", 8, 0, 0, resp, sizeof(resp)) == 0);
  TEST_ASSERT_STR_EQ(resp, "STORED");

  test_sleep_ms(1000);

  TEST_ASSERT(client_get(&c, "perm_key", val, sizeof(val), NULL) == 0);
  TEST_ASSERT_STR_EQ(val, "perm_val");

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

int test_expirations_run(void) {
  RUN_TEST(test_ttl_expiration);
  RUN_TEST(test_permanent_item);
  return 0;
}
